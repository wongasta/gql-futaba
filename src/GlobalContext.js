import React from "react";

export const GlobalContext = React.createContext({
  user: null,
  user_id: null
});