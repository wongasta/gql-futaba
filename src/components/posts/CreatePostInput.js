import React, {useRef,useContext} from "react";
import '../form_shared.css';
import { useHistory } from "react-router-dom";
import {GlobalContext} from "../../GlobalContext";
import Environment from '../../util/relayEnv';
import add_post from "../../mutations/add_post";

export default function CreatePostInput(props){
  const routerHistory = useHistory();
  const postForm = useRef(null);
  const {user,user_id} = useContext(GlobalContext);

  function handleSubmit(e){
    const form=postForm.current;
    const PostInput={
      user: user,
      title: form.querySelector('#input_title').value,
      post_content: form.querySelector('#input_content').value,
      image_url: form.querySelector('#input_image_url').value
    }
    add_post(Environment,PostInput,(data)=>{
      routerHistory.push(`/post/${data.postEdge.cursor}`);
    });
    e.preventDefault();
  }

  return (
    <form id="post_form" className={"form_container"} ref={postForm} onSubmit={handleSubmit}>
      <div className={"input_container"}>
        <div className={"labels"}>
          <label htmlFor="input_user">User</label>
          <label htmlFor="input_title">Title</label>
          <label htmlFor="input_content" className="content">Comment</label>
          <label htmlFor="input_image_url">Image</label>
        </div>
        <div className={"inputs"}>
          <input id="input_user" name="input_user" type="text" value={user_id || ""} disabled={true} />
          <input id="input_title" name="input_title" type="text" placeholder="title" required={true} />
          <textarea id="input_content" name="input_content" defaultValue={"Comment"}></textarea>
          <input id="input_image_url" name="input_image_url" type="url" placeholder="Image Link" />
        </div>
      </div>
      <div className={"submit_container"}>
        <button type="submit">Submit</button>
      </div>
    </form>
  )
}