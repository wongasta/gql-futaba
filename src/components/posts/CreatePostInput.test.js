import React from 'react';
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import CreatePostInput from './CreatePostInput';
import {GlobalContext} from "../../GlobalContext";
const add_post = require('../../mutations/add_post');
jest.mock('../../mutations/add_post', ()=>jest.fn());

describe("CreatePostInput", ()=>{

  afterAll(()=>{
    jest.clearAllMocks();
    add_post.mockRestore()
  })

  test("Construction Snapshot", ()=>{
    const Component = mount(<GlobalContext.Provider value={{user: "abc", user_id: "abc"}}><CreatePostInput /></GlobalContext.Provider>);
    expect(toJson(Component)).toMatchSnapshot();
  });

  test("Submission", ()=>{
    const Component = mount(<GlobalContext.Provider value={{user: "abc", user_id: "abc"}}><CreatePostInput /></GlobalContext.Provider>);
    Component.find("#input_title").instance().value="mock title";
    Component.find("#input_content").instance().value="mock content"
    Component.update();
    Component.find("#post_form").simulate('submit');
    expect(add_post).toHaveBeenCalledTimes(1);
    expect(add_post.mock.calls[0][1].user).toBe('abc');
    expect(add_post.mock.calls[0][1].title).toBe('mock title');
    expect(add_post.mock.calls[0][1].post_content).toBe('mock content');
  });
})