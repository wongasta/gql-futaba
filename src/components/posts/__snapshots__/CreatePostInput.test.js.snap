// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePostInput Construction Snapshot 1`] = `
<CreatePostInput>
  <form
    className="form_container"
    id="post_form"
    onSubmit={[Function]}
  >
    <div
      className="input_container"
    >
      <div
        className="labels"
      >
        <label
          htmlFor="input_user"
        >
          User
        </label>
        <label
          htmlFor="input_title"
        >
          Title
        </label>
        <label
          className="content"
          htmlFor="input_content"
        >
          Comment
        </label>
        <label
          htmlFor="input_image_url"
        >
          Image
        </label>
      </div>
      <div
        className="inputs"
      >
        <input
          disabled={true}
          id="input_user"
          name="input_user"
          type="text"
          value="abc"
        />
        <input
          id="input_title"
          name="input_title"
          placeholder="title"
          required={true}
          type="text"
        />
        <textarea
          defaultValue="Comment"
          id="input_content"
          name="input_content"
        />
        <input
          id="input_image_url"
          name="input_image_url"
          placeholder="Image Link"
          type="url"
        />
      </div>
    </div>
    <div
      className="submit_container"
    >
      <button
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</CreatePostInput>
`;
